### PR TITLE
feat: add savings charge form

### PR DESCRIPTION
**Pull request by abhi40308**
## Description
Added Add savings charge form page for adding New Saving Charge. For this purpose, created a new `Add-charge` shared component in `shared` module, which can be imported by components requiring add charge functionality.

* Add Savings Module (taking reference from community app).
* Add Savings Routing and Savings Service.
* Add new `Add-charge` component in `shared` module for adding charges, as add-charge is required by savings, loans, clients, fixed deposits and recurring deposits. This component can be reused at required places.
* Updated Add-loan-charge component to use Add-charge component.

## Related issues and discussion
#250 

## Screenshots, if any
Screenshot 2020-04-10 at 9 20 08 PM

Savings module (and loans module) is meant to be lazy loaded in clients, groups and centers. Make the following changes in clients-routing.module in order to test these components: 
Screenshot 2020-04-10 at 9 21 16 PM

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] If you have multiple commits please combine them into one commit by squashing them.

- [x] Read and understood the contribution guidelines at `web-app/.github/CONTRIBUTING.md`.

### CODE REVIEW

**Comment by abhi40308** [comment]
@abhaychawla  I've updated the pr to include some minor changes, will update again once the decision on the shared `add-charges` component is finalised.

**Comment by karantakalkar** [comment]
@abhaychawla In my opinion we shouldn't use re usable add charge form until same form is being used in 5+ components, also if this is made re usable create center/group should also be re-usable.

**Comment by abhaychawla** [comment]
@abhi40308 Lets update the PR so the component is not shared.

**Comment by abhi40308** [comment]
Done 👍

File: src/app/savings/add-savings-charge/add-savings-charge.component.html [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            <mat-form-field>
+              <mat-label>Charge time type</mat-label>
+              <mat-select formControlName="chargeTimeType">
+                <mat-option *ngFor="let chargeTime of chargeDetails.chargeTimeTypeOptions" [value]="chargeTime.id">
+                  {{ chargeTime.value }}
+                </mat-option>
+              </mat-select>
+            </mat-form-field>
+
+            <mat-form-field *ngIf="savingsChargeForm.controls['dueDate']">
```
Comment (karantakalkar) [inline src/app/savings/add-savings-charge/add-savings-charge.component.html:51]: use .contains

File: src/app/savings/add-savings-charge/add-savings-charge.component.html [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+              <input matInput [min]="minDate" [max]="maxDate" [matDatepicker]="dueDatePicker" required
+                formControlName="dueDate">
+              <mat-datepicker-toggle matSuffix [for]="dueDatePicker"></mat-datepicker-toggle>
+              <mat-datepicker #dueDatePicker></mat-datepicker>
+              <mat-error *ngIf="savingsChargeForm.controls.dueDate.hasError('required')">
+                Due for collection on is <strong>required</strong>
+              </mat-error>
+            </mat-form-field>
+
+            <mat-form-field *ngIf="savingsChargeForm.controls['feeOnMonthDay']">
```
Comment (karantakalkar) [inline src/app/savings/add-savings-charge/add-savings-charge.component.html:62]: use .contains

File: src/app/savings/add-savings-charge/add-savings-charge.component.html [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+          </div>
+
+        </div>
+
+        <mat-card-actions fxLayoutGap="5px" fxLayout="row" fxLayout.xs="column" fxLayoutAlign="center">
+          <button type="button" mat-raised-button [routerLink]="['../']">Cancel</button>
+          <button mat-raised-button color="primary" [disabled]="!savingsChargeForm.valid">Submit</button>
+        </mat-card-actions>
+
+      </mat-card-content>
```
Comment (karantakalkar) [inline src/app/savings/add-savings-charge/add-savings-charge.component.html:87]:
```suggestion
      </mat-card-content>
      
```

File: src/app/savings/add-savings-charge/add-savings-charge.component.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+/**
+ * Create Add Savings Charge component.
+ */
+@Component({
+  selector: 'mifosx-add-savings-charge',
+  templateUrl: './add-savings-charge.component.html',
+  styleUrls: ['./add-savings-charge.component.scss']
+})
+export class AddSavingsChargeComponent implements OnInit {
```
Comment (karantakalkar) [inline src/app/savings/add-savings-charge/add-savings-charge.component.ts:18]:
```suggestion
export class AddSavingsChargeComponent implements OnInit {

```

File: src/app/savings/add-savings-charge/add-savings-charge.component.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+import { Component, OnInit } from '@angular/core';
+import { FormGroup, FormBuilder, Validators, FormControl } from '@angular/forms';
+import { Router, ActivatedRoute } from '@angular/router';
+import { DatePipe } from '@angular/common';
+
+/** Custom Services */
+import { SavingsService } from '../savings.service';
+
+/**
+ * Create Add Savings Charge component.
```
Comment (karantakalkar) [inline src/app/savings/add-savings-charge/add-savings-charge.component.ts:11]:
```suggestion
 *  Add Savings Charge component.
```

File: src/app/savings/add-savings-charge/add-savings-charge.component.ts [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+  /**
+   * Creates the Savings Charge form.
+   */
+  ngOnInit() {
+    this.createSavingsChargeForm();
+    this.savingsChargeForm.controls.chargeId.valueChanges.subscribe(chargeId => {
+      this.savingsService.getChargeTemplate(chargeId).subscribe((data: any) => {
+        this.chargeDetails = data;
+
```
Comment (karantakalkar) [inline src/app/savings/add-savings-charge/add-savings-charge.component.ts:60]:
```suggestion
```

File: src/app/savings/add-savings-charge/add-savings-charge.component.ts [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        } else {
+          this.savingsChargeForm.removeControl('feeOnMonthDay');
+        }
+
+        if (data.chargeTimeType.value === 'Monthly Fee') {
+          this.savingsChargeForm.addControl('feeInterval', new FormControl(data.feeInterval, Validators.required));
+        } else {
+          this.savingsChargeForm.removeControl('feeInterval');
+        }
+
```
Comment (karantakalkar) [inline src/app/savings/add-savings-charge/add-savings-charge.component.ts:85]:
```suggestion
```

File: src/app/savings/add-savings-charge/add-savings-charge.component.ts [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    });
+    this.savingAccountId = this.route.snapshot.params['savingAccountId'];
+  }
+
+  /**
+   * Creates the Savings Charge form.
+   */
+  ngOnInit() {
+    this.createSavingsChargeForm();
+    this.savingsChargeForm.controls.chargeId.valueChanges.subscribe(chargeId => {
```
Comment (karantakalkar) [inline src/app/savings/add-savings-charge/add-savings-charge.component.ts:57]: You can move this to a new function buildDependencies()

File: src/app/savings/add-savings-charge/add-savings-charge.component.ts [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    });
+  }
+
+  /**
+   * Submits savings charge.
+   */
+  submit() {
+    const savingsCharge = this.savingsChargeForm.value;
+    savingsCharge.locale = 'en';
+
```
Comment (karantakalkar) [inline src/app/savings/add-savings-charge/add-savings-charge.component.ts:113]:
```suggestion
```

File: src/app/savings/add-savings-charge/add-savings-charge.component.ts [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+      } else {
+        const dateFormat = 'yyyy-MM-dd';
+        savingsCharge.dateFormat = dateFormat;
+        if (savingsCharge.dueDate) {
+          const prevDate = this.savingsChargeForm.value.dueDate;
+          savingsCharge.dueDate = this.datePipe.transform(prevDate, dateFormat);
+        }
+      }
+    }
+
```
Comment (karantakalkar) [inline src/app/savings/add-savings-charge/add-savings-charge.component.ts:131]:
```suggestion
```

File: src/app/savings/add-savings-charge/add-savings-charge.component.ts [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    this.savingsService.createSavingsCharge(this.savingAccountId, 'charges', savingsCharge).subscribe(res => {
+      console.log(res);
```
Comment (karantakalkar) [inline src/app/savings/add-savings-charge/add-savings-charge.component.ts:133]:
```suggestion
```

File: src/app/savings/savings-routing.module.ts [hunk 12]
```diff
@@ -7,38 +7,52 @@ import { extract } from '../core/i18n/i18n.service';
...
-                        data: { title: extract('Saving Account Actions'), routeParamBreadcrumb: 'action' },
-                        resolve: {
-                            savingAccountTransactionTemplate: SavingAccountTransactionTemplateResolver
-                        }
-                    }
-                ]
+          {
+            path: 'add-savings-charge',
+            component: AddSavingsChargeComponent,
+            data: {
```
Comment (karantakalkar) [inline src/app/savings/savings-routing.module.ts:30]: this should be in single line

File: src/app/savings/savings.module.ts [hunk 13]
```diff
@@ -9,21 +9,16 @@ import { SharedModule } from 'app/shared/shared.module';
...
-  imports: [
-    SharedModule,
-    SavingsRoutingModule
-  ],
-  declarations: [
-    SavingAccountActionsComponent,
-    SavingsAccountTransactionsComponent
-  ],
+  imports: [SharedModule, SavingsRoutingModule],
+  declarations: [SavingAccountActionsComponent, SavingsAccountTransactionsComponent, AddSavingsChargeComponent],
```
Comment (karantakalkar) [inline src/app/savings/savings.module.ts:21]: vertically arrange declarations please

Comment (karantakalkar) [reply src/app/savings/savings.module.ts:21]: and imports

**Comment by abhi40308** [comment]
@karantakalkar  i have resolved all the errors and made the logic same to community app, but there are bugs in community app code (if you select a charge, then switch to another charge and come back at the same charge again, it does not show the same fields sometimes).

**Pull request merged**